Add friends likes

MR-k0f31n opened the pull request:
Добавлены списки друзей в сущности пользователей
Добавлены списки "лайков" в сущности друзей
Добавлены эндпонты к запросам "Получить пользователя по ИД", "Получить список всех пользователей", "Получить список друзей пользователя", "Получить общих друзей с пользователем"
Добавлены эндпонты к запросам "Получить фильм по ИД", "Получить список всех фильмов", "Получить топ фильмов"
Добавить сервис добавления\Удаления в друзья, поиск общих друзей
Добавить сервис добавления\удаления лайков, получение топ фильмов 
Добавить компоненты хранения пользователей
Добавить компоненты хранения фильмов

MR-k0f31n committed:
<Start> начать рефакторинг, добавить сервисы, добавить обработчик ошибок

MR-k0f31n committed:
<Add> Добавить собственную валидацию на дату релиза фильма. Добавить в сущности пользователей список друзей. Добавить в сущности фильма список лайков. Добавить сервисы фильмов и пользователей. Добавить компоненты хранения пользователей и фильмов.

TomattmanYandex commented on src/main/java/ru/yandex/practicum/filmorate/controller/ErrorHandler.java:
```diff
@@ -0,0 +1,52 @@
+
+    @ExceptionHandler
+    @ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)
+    public Map<String, String> handleNullPointerException(final NullPointerException exception) {
```
хороший момент, но такие обработчики делать не обязательно, у спринга есть встроенные обработчики ошибок, и они делают примерно то же что и этот метод

MR-k0f31n replied:
Не очень понял, тут же получается как нас и учили либо отдельный респонс делать либо можно обработать так и указать какой именно код сервер должен ответь. я так и сделал

TomattmanYandex commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,70 @@
+    private final FilmService service;
+
+    @Autowired
+    public FilmController (FilmService service) {
```
можно добавить аннотацию RequiredArgsConstructor Она создаст точно такой же конструктор очень удобно для дальнейшего расширения класса, при добавлении зависимостей не нужно будет дополнительно менять конструктор

MR-k0f31n replied:
ооо тут были отдельные пляски, я тупой бошка подай пирожка в контроллере использовал разные переменные пути и имена переменных, а ругал во всем компоненты и сервисы, согласен, заменю.

TomattmanYandex commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -24,7 +24,9 @@ public class Film {
     private final LocalDate releaseDate;
     @Positive(message = "должно быть больше 0")
     private final Integer duration;
+    private final Set<Integer> userLike = new HashSet<>();
```
отлично что использован set а не list

MR-k0f31n replied:
изначально хотел сделать List с проверкой есть нету и всего один метод, если пользователя нет значит лайк если пользователь уже лайкал - значит ремув лайк, но не знаю насколько это верно было бы, поэтому через сет бороться начал

TomattmanYandex commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,63 @@
+@Service
+@RequiredArgsConstructor
+@Slf4j
+public class FilmService {
```
В этот класс хорошо было бы добавить зависимость от UserService или UserStorage и при лайках проверять, есть ли такой юзер

MR-k0f31n replied:
у меня была идея еще и фильмы юзеру записывать, а потом подумал, а надо до ли?

MR-k0f31n replied:
если в дальнейшем конечно поставять задачу записывать и выводить список понравившегося пользоваетлю

TomattmanYandex replied:
пока не надо усложнять, как доберетесь до баз данных, все изменится, и придётся все равно переделывать

TomattmanYandex commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -0,0 +1,94 @@
+    public List<User> commonFriend (int userId, int otherId) {
+        log.trace("Попытка получить список общих друзей");
+        Set<Integer> commonFriendsId = userStorage.getUserById(userId).getFriends()
+                .stream().filter(userStorage.getUserById(otherId).getFriends()::contains)
```
лучше вынести получение второго пользователя из стрима, явная оптимизация не надо будет нессколько раз пытаться получить одного и того же пользователя

MR-k0f31n replied:
Тоже думал что и конструкция громадная и явные проблемы, но это первые пробы стримов лямбд =) как хорошо что у меня ревью ткнет куда нужно

TomattmanYandex replied:
не сама конструкция, только то, что получаешь из хранилища в самом стриме, если у тебя будет у кого то 10 друзей, то второго пользователя, одного и того же, получим 10 раз

TomattmanYandex reviewed (changes requested):
Не понятно почему контроллеры были перемещены в другой пакет, но ладно
Отличная работа, пара небольших изменений и будет все ок

MR-k0f31n committed:
<FIX> Исправить - проверку наличие пользователя при постановке лайка.
Исправить - оптимизация получения списка общих друзей.
Исправить - автоматическа анотация добовления зависимостей в контруктор.

MR-k0f31n reviewed (commented):
<FIX> Исправить - проверку наличие пользователя при постановке лайка.
Исправить - оптимизация получения списка общих друзей.
Исправить - автоматическа анотация добовления зависимостей в контруктор.

TomattmanYandex reviewed (approved):
Принято
